File: MinStack_design.py
"""
Author: Srinidhi Bhat 
Time Complexity : O(1) - All Operation
Space Complexity : O(N) - In worst case all can be push operations 
Did this code successfully run on Leetcode : Yes - faster than 99.7% and memory less than 10%

Any problem you faced while coding this :No

Core Logic - 
Remaining Stack operations are same
To get Min in O(1) - Always maintain the minimum at every block as a tuple
the 1st position of tuple will be element and 2nd position will be the minimum found till that time
"""
class MinStack:

    def __init__(self):
        self.my_stack = []

    def push(self, x: int) -> None:
        #While pushing, maintain the structure - (element,minimum)
        if len(self.my_stack) == 0:
            self.my_stack.append((x,x))
        else:
            curr_min = self.my_stack[-1][1]
            if x < curr_min:
                curr_min = x
            self.my_stack.append((x,curr_min))

    def pop(self) -> None:
        return self.my_stack.pop()
    
    def top(self) -> int:
        return self.my_stack[-1][0] #since element is in 0th position

    def getMin(self) -> int:
        return self.my_stack[-1][1] #since minimum is in the 1st position

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: MinStack_Design.java
//T.C: here everything is O(1) - push , pop, min
//S.C: (2(n)) ==> O(n)

class MinStack {
    Stack<Integer> mainSt;
    Stack<Integer> minSt;
    int min;
    public MinStack() {
        mainSt = new Stack<>();
        minSt = new Stack<>();
        min = Integer.MAX_VALUE; // adding to min initially
        minSt.add(min); 
        
    }
    
    public void push(int val) {
         mainSt.push(val);
        if(val < minSt.peek()){ // check val with the min value 
            minSt.push(val);
            min = val; // update min value
        } else {
            minSt.push(min);
        }
         
    }
    
    public void pop() {
         mainSt.pop();
         minSt.pop();
         min = minSt.peek(); // update the min value
    }
    
    public int top() {
        return mainSt.peek();
    }
    
    public int getMin() {
         return minSt.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */